### ISSUE TITLE
[question] Package immutability and user/channel inconsistencies?

### ISSUE THREAD

**Issue by gsemet** [conan-io/conan#14207]
### What is your question?

Hello.

I think I can understand why is that so important to have package immutability on Conan 2, but do you have a clear article or part in the documentation that explain in detail why Conan really wanted this?

I give a quick summary of my use case:
- I have a set of various inside packages, some open source package, mostly company internal conan 1 packages
- user/channel are really a mess, without clear convention. So, **first question,** is there recommendation for package naming inside a company that may have several departments working on different subject, sometime wanting to share some packages, sometime having more "business oriented" custom packages? I will propose a convention for official packages user/channel to follow `<mainprogram>_<department>_<platform>/channel`, so one department can "easy" import packages from another one in their repositories and there won't be conflicts neither a single folder will all packages. But is there better convention ?
- so, today we have a mess with user/channel on existing packages, and I need to reuse some in my new projects. And I would like it to be cleaner.

Importing some package (ie copy into our working repository) directly will result in copying the mess, but even using "conan copy" in Conan 1 would not save the dependency problem. I need basically a way to tell "everything that were in "baduser/badchannel", once imported, you will find them into "gooduser/goodchannel". I see no use to change the package name or version.

As I understand, the philosophy of Conan2 is to have immutable package, and in this case it will help because all the requirements or tools_requirements will "stay" the same without adaptation. And I understand in theory I would have to "create new package" in order to fit my old packages into the new structure.

So my $1000 question is: is there a way to "import" an existing package by copying everything it contains (recipe, name, version, binaries, ...) but just change its user/channel (that would be the equivalent of "conan copy" but also "replace" all it dependencies reference to my "user/channel"? Even if this mean to rebuild from source or change the package_id/revision_id, I think this is acceptable. The goal is to control better the user/channel section in the new structure.

Or I have to do a reimplementation from scratch, recopying the whole original recipe ? Of course, I know I need to do it for each transitivie dependencies, so if the "import recipe" is pretty easy to implement, I guess this is some i can manually do for each of its depedencies (there are not so many, most of our packages share the same single dependency).

Thanks for your anwser
Gaetan

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @gsemet 

I think the main ideas are:

- Basically drop the user/channel as a way of qualifying things. It is useful to identify packages with the name of the team, specially in very large companies. (In 2.0, the channel can be dropped and use only the user)
- The way to control sets of dependencies if by "promoting" them to different server repositories. This is a well known good practice in package management and devops. So packages and their dependencies are never renamed or re-created from scratch (also a bad practice), but instead, they are copied to a different repo. For example there could be a "develop", a "testing" and a "prod" server repos. When a set of dependencies (a package-list in Conan, for example) from ``develop`` is looking good, it can be copied to the ``testing`` repo, making them available for the testing team. No need to rename them, just copying them and resolving dependencies against the ``testing`` repo will make it. Same later for ``prod``.
- This approach completely avoids conditional or variable ``requires`` and any change in packages at all, so complete immutability and also ensuring that things don't need to be rebuilt from scratch.

The docs for "CI" are still pending, we are starting to illustrate some of these concepts in places like https://docs.conan.io/2/examples/commands/pkglists.html#downloading-from-one-remote-and-uploading-to-a-different-remote

**Comment by gsemet** [comment]
Hello. Thanks for your answer. I understand this "promotion through repository copy", this is fine, I am fully using it between the "testing" repository and the "official release" repository.

But I want to kind of reorganize my dependencies. I wish to be able to store pretty much everything that is needed to build, test, release the huge platform I am working on (several hundred depedencies, ...) and if needed, reorganize the dependencies.

For example, let's say my component C1 needs 2 libs L1 and L2, and tools T1 and T2. Some tools are opensource open, stored on conan center, we have a cache on a dedicated, open repository on our internal Artifactory. Some libs (say L1) comes from another department and they stored in a weird user/channel configuration.  Now we can have interdependencies in these packages, saying for instance L2/1.0.0@baduser/badchannel needs a lib L3/2.0.0@baduser/badchannel.

I want to store all these internal dependencies in a single repository also to avoid nightmare in permissions, because some of these dependencies shall not be available to anyone else. So I plan in having an "import" process to basically recopy these messy dependencies (all but conancenter cache) in our "clean" repository accessible to every of our developers. But I would like to reorganize the user/channel of each lib and there self.requires, self.tools_requires and so on,... without changing any other thing. This is a breakage in the package immutability, and I would be ok to "repackage" them into different package, I am just looking for an easy way to do it.

In short: I want to be able to take a package A and repackage it and some of its dependencies into another user/channel, and I am looking for the easiest way of doing it.

PS: we followed the convention to use "teamname/stable" for the user/channel convention, so they might already live like this in our repository, but it one day we have a conflict we can have import problem.

Regards,

**Comment by memsharded** [comment]
Thanks for the clarifications, that helps a lot to understand what you are trying to achieve and your previous comments.

> So my $1000 question is: is there a way to "import" an existing package by copying everything it contains (recipe, name, version, binaries, ...) but just change its user/channel (that would be the equivalent of "conan copy" but also "replace" all it dependencies reference to my "user/channel"?

So, I am afraid there isn't a builtin or documented way. It depends on how package recipes are implemented, if they parameterize their requires as ``self.requires(f"dep/version@{self.user}/{self.channel}")``, I guess that some forced user/channel rename could make things work, even if it is not recommended. This might be doable for example in the server side, by forcing a copy to another location controlling the folders that match the user/channel parts. Have you considered this possibility?

If the recipe hardcodes ``self.requires("dep/versioon@myuser/mychannel")`` then, it will be completely impossible without modifying the recipe and rebuilding from source.

**Comment by gsemet** [comment]
ok I understand. I have dealt with packages that does `dep/version@{self.user}/{self.channel}` because the original authors thought it was easier and more flexible, but this was really a nightmare to use.

Now we have pretty much in the "hardcoded" way (and this will be compatible with Conan 2 I hope). If the use case I want to do is not possible, so you have a recommendation for the "user/channel" a given team can release?

I push to set in place the convention "<program>_<teams>_<increment>/stable" for in-production binaires, `<program>_team` for reusable tools, so that they can be easily "promoted" to another repository, hoping to minimize the risk of conflicts.

**Comment by memsharded** [comment]
Closing this question as answered back then, but please don't hesitate to comment or create new tickets for any further question or issue, thanks for the feedback!
